Feat/undo actions

acc-erick-rosas opened the pull request.

acc-erick-rosas committed:
Instant search  + minor fixes

acc-erick-rosas committed:
Included Snack Bar Component to Undo changes

catfireparty commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -18,3 +18,43 @@ describe('When: I use the reading list feature', () => {
+        await openReadingList.click();
+      }
+    }
+    await browser.sleep(600);
```
Do you need to use `browser.sleep` quite so liberally? Are there better ways to wait for things to be available?

catfireparty commented on apps/okreads-e2e/src/specs/search-books.spec.ts:
```diff
@@ -16,12 +16,43 @@ describe('When: Use the search feature', () => {
+    await browser.sleep(600);
+
+    const bookItemList = await $$('[data-testing="btn-add-book"]:enabled');
+    if(bookItemList) {
```
if the `bookItemList` doesn't exist then this test will still pass. Is that what you meant to do?

catfireparty commented on libs/api/books/src/lib/books.service.ts:
```diff
@@ -15,20 +15,25 @@ export class BooksService {
-          });
+        map((resp) => {
+          if (resp.data.totalItems > 0) {
+            console.log('SOOOOO OH AHHHHHHH :: ', resp);
```
🤨

catfireparty commented on libs/api/books/src/lib/reading-list.controller.ts:
```diff
@@ -16,8 +16,18 @@ export class ReadingListController {
     return await this.readingList.addBook(item);
   }
 
+  @Delete('/reading-list/:id')
```
I'm surprised the API still works with two routes exactly the same.

You don't need to do this.

catfireparty commented on libs/books/feature/src/lib/book-search/book-search.component.ts:
```diff
@@ -45,18 +50,35 @@ export class BookSearchComponent implements OnInit {
     }
   }
+
+  instantSearch(searchString) {
```
Why is this code in here?

acc-erick-rosas replied:
Sorry, seems like I branched from instant search instead of code review, I'll remove the code

catfireparty commented on libs/books/feature/src/lib/reading-list/reading-list.component.ts:
```diff
@@ -10,9 +12,20 @@ import { getReadingList, removeFromReadingList } from '@tmo/books/data-access';
 
   removeFromReadingList(item) {
     this.store.dispatch(removeFromReadingList({ item }));
+    this.snackBar.openFromComponent(
```
It looks like this would open simultaneously with the action that it needs to undo. Could these ever get out of sync?

Shouldn't this show when the action has been done?

acc-erick-rosas replied:
Oh, about that, no, It cant overlap or open simultaneously, as the snack bar is bound to the click on the buttons and not action or reducer, so we can dispatch the action without worrying about  the snack bar triggering again.

catfireparty reviewed (commented):
I think you need to take another look at your approach here. Use the actions and effects more. You probably don't need to create a new snackbar component at all.

You're also lacking tests generally and the e2e test is kind of a mess. You could use a lot less code to do the same thing with fewer `sleep` calls.

acc-erick-rosas committed:
Updated e2e and controller

acc-erick-rosas commented:
@jdpearce I updated the e2e test, removing all the sleeps, I also removed the insta-search code, the extra end-points for the controller and other stuff. I had to leave the component approach for the snackbar, as I tried to implement the more "direct" way you mention, but the e2e test got more complicated than it is now, and even I couldn't make it work as the snackbar kept getting out of scope. I guess this approach shouldn't be a problem, as it is suggested in the official documentation :)
